### PR TITLE
[ews-build.webkit.org] Land from branch aliases

### PR DESCRIPTION
**Pull request by JonWBedard**
#### c8afe4b6ecf05585639c5758ac08641c520c3cb1
<pre>
[ews-build.webkit.org] Land from branch aliases
<a href="https://bugs.webkit.org/show_bug.cgi?id=247979">https://bugs.webkit.org/show_bug.cgi?id=247979</a>
rdar://102419945

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/factories.py:
(MergeQueueFactoryBase.__init__): Add MapBranchAlias step.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps): Add &apos;map-branch-alias&apos;.
* Tools/CISupport/ews-build/steps.py:
(MapBranchAlias):
(MapBranchAlias.__init__):
(MapBranchAlias.start): List all branches which contain the specified branch name.
(MapBranchAlias.getResultSummary):
(MapBranchAlias.evaluateCommand): If the specified branch is a production branch,
use it. Otherwise, prefer a production branch with contains the specified branch.
(MapBranchAlias.doStepIf): Map branch alias if the specified branch is &apos;main&apos;.
(MapBranchAlias.hideStepIf): Hide step if we aren&apos;t doing it.
* Tools/CISupport/ews-build/steps_unittest.py:
(TestMapBranchAlias):

Canonical link: <a href="https://commits.webkit.org/256756@main">https://commits.webkit.org/256756@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b83f16625fc43aa22c5afaea0877c648046cdbcf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96755 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29849 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/106283 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6234 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/34754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89142 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/102989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102430 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/83370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/57 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/95709 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/52 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/4839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/83370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2251 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/1278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

File: Tools/CISupport/ews-build/steps.py [hunk 1]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
...
+class MapBranchAlias(shell.ShellCommand):
+    name = 'map-branch-alias'
+    haltOnFailure = False
+    flunkOnFailure = True
+    DEV_BRANCHES = re.compile(r'.*[(eng)(dev)(bug)]/.+')
+    PROD_BRANCHES = re.compile(r'\S+-[\d+\.]+-branch')
+
+    def __init__(self, **kwargs):
+        self.summary = ''
+        super(MapBranchAlias, self).__init__(logEnviron=False, **kwargs)
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: do we want to add a `timeout` here?

File: Tools/CISupport/ews-build/steps.py [hunk 2]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
...
+
+    def start(self, BufferLogObserverClass=logobserver.BufferLogObserver):
+        base_ref = self.getProperty('github.base.ref', DEFAULT_BRANCH)
+        remote = self.getProperty('remote', DEFAULT_REMOTE)
+
+        self.command = [
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: can make it single line for better readability

File: Tools/CISupport/ews-build/steps.py [hunk 3]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
...
+                    break
+
+        self.summary = f"'{branch}' is the prevailing alias"
+        self.setProperty('github.base.ref', branch)
+        return rc
+
+    def doStepIf(self, step):
+        if not self.getProperty('github.number'):
+            return False
+        remote = self.getProperty('remote', DEFAULT_REMOTE)
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Is it intentional that `remote` is unused?

File: Tools/CISupport/ews-build/steps.py [hunk 4]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
...
+                continue
+            candidate = line.split('/', 2)[-1]
+            if self.DEV_BRANCHES.match(candidate):
+                continue
+            aliases.add(candidate)
+
+        if DEFAULT_BRANCH in aliases:
+            branch = DEFAULT_BRANCH
+        if branch != DEFAULT_BRANCH and self.DEV_BRANCHES.match(branch):
+            for candidate in aliases:
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: `candidate` variable is re-used to represent multiple different things. Probably better to name it alias here, e.g.: `for alias in aliases:`

File: Tools/CISupport/ews-build/steps.py [hunk 5]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
...
+                branch = candidate
+                break
+        if branch != DEFAULT_BRANCH and not self.PROD_BRANCHES.match(branch):
+            for candidate in aliases:
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Ditto

File: Tools/CISupport/ews-build/steps.py [hunk 6]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
...
+            if self.DEV_BRANCHES.match(candidate):
+                continue
+            aliases.add(candidate)
+
+        if DEFAULT_BRANCH in aliases:
+            branch = DEFAULT_BRANCH
+        if branch != DEFAULT_BRANCH and self.DEV_BRANCHES.match(branch):
+            for candidate in aliases:
+                branch = candidate
+                break
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: why are we running a loop just to break it in first iteration, we can just refer to aliases[0] if aliases isn't None.

File: Tools/CISupport/ews-build/steps.py [hunk 7]
```diff
@@ -4860,6 +4860,85 @@ def hideStepIf(self, results, step):
         return not self.doStepIf(step)
 
 
+class MapBranchAlias(shell.ShellCommand):
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py:4886]: It would be good to add a comment here describing what this step is doing (since it's not obvious)

**Review by aj062 — APPROVED** [review]

File: Tools/CISupport/ews-build/steps.py (hunk 6 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/steps.py]: It's a set so we can't index...I guess `next(aliases)` would have the same effect?

Comment (aj062) [reply Tools/CISupport/ews-build/steps.py]: maybe `list(aliases)[0]`

File: Tools/CISupport/ews-build/steps.py (hunk 3 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/steps.py]: Was a hold-over from a previous implementation.

**Pull request merged**